### PR TITLE
Add Class-Aware Sampler

### PR DESCRIPTION
**Pull request by yeyun111**
Implemented Class-Aware Sampling: https://arxiv.org/abs/1512.05830. Li Shen, Zhouchen Lin, Qingming Huang, Relay Backpropagation for Effective Learning of Deep Convolutional Neural Networks, ECCV 2016

Use example:
```Python
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
from torch.utils.data.sampler import ClassAwareSampler

train_set = datasets.ImageFolder(
    '/path/to/data/train',
    transforms.Compose([
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ]))
train_loader = DataLoader(train_set, batch_size=4, sampler=ClassAwareSampler(train_set))
```

### CODE REVIEW

**Comment by YanWang2014** [comment]
Great! I am just looking for this.

**Comment by weiyangfb** [comment]
thoughts on this PR? @fmassa @SsnL

File: torch/utils/data/sampler.py [hunk 1]
```diff
@@ -90,6 +123,39 @@ def __len__(self):
...
+    Implemented Class-Aware Sampling: https://arxiv.org/abs/1512.05830
+    Li Shen, Zhouchen Lin, Qingming Huang, Relay Backpropagation for Effective
+    Learning of Deep Convolutional Neural Networks, ECCV 2016
+    By default num_samples equals to number of samples in the largest class
+    multiplied by num of classes such that all samples can be sampled.
+    """
+
+    def __init__(self, data_source, num_samples=0):
+        self.data_source = data_source
+        n_cls = len(data_source.classes)
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: This makes the assumption that `data_source` has a `classes` field, which is not in general valid

File: torch/utils/data/sampler.py [hunk 2]
```diff
@@ -90,6 +123,39 @@ def __len__(self):
...
+        self.class_iter = RandomCycleIter(range(n_cls))
+        cls_data_list = [list() for _ in range(n_cls)]
+        for i, (_, label) in enumerate(data_source.imgs):
```
Comment (fmassa) [inline torch/utils/data/sampler.py]: Same comment here, `imgs` is not generally an attribute of a `Dataset`.

**Review by fmassa — CHANGES_REQUESTED** [review]
Hi,

Thanks for the PR and sorry for the delay in reviewing it.

While this is a nice addition, it makes assumptions that do not hold for general `Dataset` classes.
For this reason, I think it might be better to submit this PR to [`torchcontrib`](https://github.com/pytorch/contrib).

Thanks!

**Comment by EaglesKK** [comment]
hi, @fmassa @frombeijingwithlove 
I'm new here, and I was trying to use the ClassAwareSampler, but after I tried `from torchcontrib import ClassAwareSampler` it didn't work, could you tell me how to use it properly? Thanks

**Pull request closed without merging**